core/src/main/java/org/elasticsearch/cluster/routing/allocation/UnassignedShardDecision.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.UnassignedInfo.AllocationStatus;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision.Type;
+import org.elasticsearch.common.Nullable;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Represents the allocation decision by an allocator for an unassigned shard.
+ */
+public class UnassignedShardDecision {
+    /** a constant representing a shard decision where no decision was taken */
+    public static final UnassignedShardDecision DECISION_NOT_TAKEN =
+        new UnassignedShardDecision(false, null, null, null, null, null, null);
+
+    private final boolean decisionTaken;
+    @Nullable
+    private final Decision finalDecision;
+    @Nullable
+    private final AllocationStatus allocationStatus;
+    @Nullable
+    private final String finalExplanation;
+    @Nullable
+    private final String assignedNodeId;
+    @Nullable
+    private final String allocationId;
+    @Nullable
+    private final Map<String, Decision> nodeDecisions;
+
+    private UnassignedShardDecision(boolean decisionTaken,
+                                    Decision finalDecision,
+                                    AllocationStatus allocationStatus,
+                                    String finalExplanation,
+                                    String assignedNodeId,
+                                    String allocationId,
+                                    Map<String, Decision> nodeDecisions) {
+        assert finalDecision != null || decisionTaken == false :
```

Comment:
I agree, I'll make the change

On Wed, Sep 7, 2016 at 10:23 AM, Boaz Leskes <notifications@github.com>
wrote:

> In core/src/main/java/org/elasticsearch/cluster/routing/allocation/
> UnassignedShardDecision.java
> <https://github.com/elastic/elasticsearch/pull/20347#discussion_r77831502>
> :
>
> > +    private final String finalExplanation;
> > +    @Nullable
> > +    private final String assignedNodeId;
> > +    @Nullable
> > +    private final String allocationId;
> > +    @Nullable
> > +    private final Map<String, Decision> nodeDecisions;
> > +
> > +    private UnassignedShardDecision(boolean decisionTaken,
> > +                                    Decision finalDecision,
> > +                                    AllocationStatus allocationStatus,
> > +                                    String finalExplanation,
> > +                                    String assignedNodeId,
> > +                                    String allocationId,
> > +                                    Map<String, Decision> nodeDecisions) {
> > +        assert finalDecision != null || decisionTaken == false :
>
> I think the method is good but what don't need internal state imo
> … <#m_-2274511261399926767_>
> On 7 sep. 2016 3:27 PM +0200, Ali Beyad ***@***.***>, wrote: In
> core/src/main/java/org/elasticsearch/cluster/routing/allocation/UnassignedShardDecision.java
> (#20347 (comment)
> <https://github.com/elastic/elasticsearch/pull/20347#discussion_r77820791>):
> > + private final String finalExplanation; > + @Nullable
> <https://github.com/Nullable> > + private final String assignedNodeId; >
> + @Nullable <https://github.com/Nullable> > + private final String
> allocationId; > + @Nullable <https://github.com/Nullable> > + private
> final Map<String, Decision> nodeDecisions; > + > + private
> UnassignedShardDecision(boolean decisionTaken, > + Decision
> finalDecision, > + AllocationStatus allocationStatus, > + String
> finalExplanation, > + String assignedNodeId, > + String allocationId, > +
> Map<String, Decision> nodeDecisions) { > + assert finalDecision != null ||
> decisionTaken == false : That is true, decisionTaken is not strictly
> necessary as one could just test if (getFinalDecision() != null) to know if
> the object has meaningful data instead of if (isDecisionTaken()) — You are
> receiving this because you commented. Reply to this email directly, view it
> on GitHub (https://github.com/elastic/elasticsearch/pull/20347/files/
> f978fa2cdd585e776ce4c47f002c2243249827f8#r77820791), or mute the thread (
> https://github.com/notifications/unsubscribe-auth/AA9bJ4t0IOBeFUWe-
> bFCfw4DxTitlnASks5qnruhgaJpZM4J2Nch).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/20347/files/f978fa2cdd585e776ce4c47f002c2243249827f8#r77831502>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABjkQacu81aiXzvSSgDdcM5yx6CA8Xzuks5qnsjegaJpZM4J2Nch>
> .
>